algebra/ring.lean

```diff
@@ -15,6 +15,85 @@ section
   (left_distrib n 1 1).trans (by simp)
 end
 
+section units
+  variables [semiring α]
+  variable (α)
+
+  structure invertible :=
+    (val : α)
+    (inv : α)
+    (property : val * inv = 1 ∧ inv * val = 1)
+
+  variables {a b c : invertible α}
+  variable {α}
+
+  def ext (HAB : a.val = b.val) : a = b :=
+  begin
+    cases a,
+    cases b,
+    dsimp at HAB,
+    congr,
+    exact HAB,
+    exact calc
+        inv = inv * (val_1 * inv_1) : by rw [property_1.1, mul_one]
+        ... = (inv * val_1) * inv_1 : by rw [mul_assoc]
+        ... = (inv * val) * inv_1   : by rw [HAB]
+        ... = inv_1                 : by rw [property.2, one_mul]
+  end
+
+  lemma mul_four {a b c d : α} :
+  (a * b) * (c * d) = a * (b * c) * d := by simp
+
+  def mul (a b : invertible α) : invertible α :=
+  {
+    val := a.val * b.val,
+    inv := b.inv * a.inv,
+    property := ⟨by rw [mul_four,b.property.1,mul_one,a.property.1],
+                by rw [mul_four,a.property.2,mul_one,b.property.2]⟩
+  }
```

Comment:
Style: You should format structures like this:
```
{ a  := proof,
  b  := proof,
  cd := proof }
```

Reply:
Also spaces after commands: `rw [mul_four, b.property, ...]`